File: lib/factorix/serializer.rb
# frozen_string_literal: true

module Factorix
  # Serialize data to binary format
  class Serializer
    # Create a new Serializer instance
    #
    # @param stream [IO] An IO-like object that responds to #write
    # @raise [ArgumentError] If the stream doesn't respond to #write
    def initialize(stream)
      raise ArgumentError, "can't read from the given argument" unless stream.respond_to?(:write)

      @stream = stream
    end

    # Write raw bytes to the stream
    #
    # @param data [String] Binary data to write
    # @raise [ArgumentError] If data is nil
    # @return [void]
    def write_bytes(data)
      raise ArgumentError if data.nil?
      return if data.empty?

      @stream.write(data)
    end

    # Write an unsigned 8-bit integer
    #
    # @param uint8 [Integer] 8-bit unsigned integer
    # @return [void]
    def write_u8(uint8) = write_bytes([uint8].pack("C"))

    # Write an unsigned 16-bit integer
    #
    # @param uint16 [Integer] 16-bit unsigned integer
    # @return [void]
    def write_u16(uint16) = write_bytes([uint16].pack("v"))

    # Write an unsigned 32-bit integer
    #
    # @param uint32 [Integer] 32-bit unsigned integer
    # @return [void]
    def write_u32(uint32) = write_bytes([uint32].pack("V"))

    # Write a space-optimized 16-bit unsigned integer
    # https://wiki.factorio.com/Data_types#Space_Optimized
    #
    # @param uint16 [Integer] 16-bit unsigned integer
    # @return [void]
    def write_optim_u16(uint16)
      if uint16 < 0xFF
        write_u8(uint16 & 0xFF)
      else
        write_u8(0xFF)
        write_u16(uint16)
      end
    end

    # Write a space-optimized 32-bit unsigned integer
    # https://wiki.factorio.com/Data_types#Space_Optimized
    #
    # @param uint32 [Integer] 32-bit unsigned integer
    # @return [void]
    def write_optim_u32(uint32)
      if uint32 < 0xFF
        write_u8(uint32 & 0xFF)
      else
        write_u8(0xFF)
        write_u32(uint32)
      end
    end

    # Write a boolean value
    #
    # @param bool [Boolean] Boolean value
    # @return [void]
    def write_bool(bool) = write_u8(bool ? 0x01 : 0x00)

    # Write a string
    #
    # @param str [String] String to write
    # @return [void]
    def write_str(str)
      write_optim_u32(str.length)
      write_bytes(str.b)
    end

    # Write a string property
    # https://wiki.factorio.com/Property_tree#String
    #
    # @param str [String] String to write
    # @return [void]
    def write_str_property(str)
      if str.empty?
        write_bool(true)
      else
        write_bool(false)
        write_str(str)
      end
    end

    # Write a double-precision floating point number
    # https://wiki.factorio.com/Property_tree#Number
    #
    # @param dbl [Float] Double-precision floating point number
    # @return [void]
    def write_double(dbl) = write_bytes([dbl].pack("d"))

    # Write a Version64 object
    #
    # @param v64 [Version64] Version64 object
    # @return [void]
    def write_version64(v64) = v64.to_a.each {|u16| write_u16(u16) }

    # Write a Version24 object
    #
    # @param v24 [Version24] Version24 object
    # @return [void]
    def write_version24(v24) = v24.to_a.each {|u16| write_optim_u16(u16) }

    # Write a list
    # https://wiki.factorio.com/Property_tree#List
    #
    # @param list [Array] List of objects
    # @return [void]
    def write_list(list)
      write_optim_u32(list.size)
      list.each {|e| write_property_tree(e) }
    end

    # Write a dictionary
    # https://wiki.factorio.com/Property_tree#Dictionary
    #
    # @param dict [Hash] Dictionary of key-value pairs
    # @return [void]
    def write_dictionary(dict)
      write_u32(dict.size)
      dict.each do |(key, value)|
        write_str_property(key)
        write_property_tree(value)
      end
    end

    # Write a property tree
    #
    # @param obj [Object] Object to write
    # @raise [UnknownPropertyType] If the object type is not supported
    # @return [void]
    def write_property_tree(obj)
      case obj
      in true | false => bool
        write_u8(1)
        write_bool(false)
        write_bool(bool)
      in Float => dbl
        write_u8(2)
        write_bool(false)
        write_double(dbl)
      in String => str
        case str
        when /\Argba:(?<r>\h{2})(?<g>\h{2})(?<b>\h{2})(?<a>\h{2})\z/
          # convert "rgba:RRGGBBAA" to {"r": RR, "g": GG, "b": BB, "a": AA }"
          write_u8(5)
          write_bool(false)
          write_dictionary(%w[r g b a].each_with_object({}) {|k, dict| dict[k] = $~[k].to_i(16) / 255.0 })
        else
          write_u8(3)
          write_bool(false)
          write_str_property(str)
        end
      in Array => list
        write_u8(4)
        write_bool(false)
        write_list(list)
      in Hash => dict
        write_u8(5)
        write_bool(false)
        write_dictionary(dict)
      else
        raise Factorix::UnknownPropertyType, obj.class
      end
    end
  end
end

File: internal/serdes/serializer.go
package serdes

import (
	"encoding/binary"
	"fmt"
	"io"
	"math"
	"unicode/utf8"

	"github.com/sakuro/factorix/internal/mod"
)

// Serializer writes Factorio's binary format to a stream.
type Serializer struct {
	w io.Writer
}

// NewSerializer wraps w for writing.
func NewSerializer(w io.Writer) *Serializer {
	return &Serializer{w: w}
}

// WriteBytes writes raw bytes.
func (s *Serializer) WriteBytes(p []byte) error {
	_, err := s.w.Write(p)
	return err
}

// WriteU8 writes an unsigned 8-bit integer.
func (s *Serializer) WriteU8(v uint8) error {
	return s.WriteBytes([]byte{v})
}

// WriteU16 writes a little-endian unsigned 16-bit integer.
func (s *Serializer) WriteU16(v uint16) error {
	return s.WriteBytes(binary.LittleEndian.AppendUint16(nil, v))
}

// WriteU32 writes a little-endian unsigned 32-bit integer.
func (s *Serializer) WriteU32(v uint32) error {
	return s.WriteBytes(binary.LittleEndian.AppendUint32(nil, v))
}

// WriteOptimU16 writes a space-optimized 16-bit unsigned integer.
//
// See https://wiki.factorio.com/Data_types#Space_Optimized
func (s *Serializer) WriteOptimU16(v uint16) error {
	if v < 0xFF {
		return s.WriteU8(uint8(v))
	}
	if err := s.WriteU8(0xFF); err != nil {
		return err
	}
	return s.WriteU16(v)
}

// WriteOptimU32 writes a space-optimized 32-bit unsigned integer.
//
// See https://wiki.factorio.com/Data_types#Space_Optimized
func (s *Serializer) WriteOptimU32(v uint32) error {
	if v < 0xFF {
		return s.WriteU8(uint8(v))
	}
	if err := s.WriteU8(0xFF); err != nil {
		return err
	}
	return s.WriteU32(v)
}

// WriteBool writes a boolean value.
func (s *Serializer) WriteBool(v bool) error {
	if v {
		return s.WriteU8(0x01)
	}
	return s.WriteU8(0x00)
}

// WriteStr writes a length-prefixed string. The string must be valid UTF-8.
func (s *Serializer) WriteStr(str string) error {
	if !utf8.ValidString(str) {
		return fmt.Errorf("%w: %q", ErrInvalidUTF8, str)
	}
	if err := s.WriteOptimU32(uint32(len(str))); err != nil {
		return err
	}
	return s.WriteBytes([]byte(str))
}

// WriteStrProperty writes a string preceded by an "is empty" flag.
//
// See https://wiki.factorio.com/Property_tree#String
func (s *Serializer) WriteStrProperty(str string) error {
	if str == "" {
		return s.WriteBool(true)
	}
	if err := s.WriteBool(false); err != nil {
		return err
	}
	return s.WriteStr(str)
}

// WriteDouble writes a little-endian double-precision floating point number.
func (s *Serializer) WriteDouble(v float64) error {
	return s.WriteBytes(binary.LittleEndian.AppendUint64(nil, math.Float64bits(v)))
}

// WriteLong writes a little-endian signed 64-bit integer.
func (s *Serializer) WriteLong(v int64) error {
	return s.WriteUnsignedLong(uint64(v))
}

// WriteUnsignedLong writes a little-endian unsigned 64-bit integer.
func (s *Serializer) WriteUnsignedLong(v uint64) error {
	return s.WriteBytes(binary.LittleEndian.AppendUint64(nil, v))
}

// WriteGameVersion writes a GameVersion (4 x u16).
func (s *Serializer) WriteGameVersion(v mod.GameVersion) error {
	for _, u := range [...]uint16{v.Major, v.Minor, v.Patch, v.Build} {
		if err := s.WriteU16(u); err != nil {
			return err
		}
	}
	return nil
}

// WriteMODVersion writes a MODVersion (3 x optim u16).
func (s *Serializer) WriteMODVersion(v mod.MODVersion) error {
	for _, u := range [...]uint16{v.Major, v.Minor, v.Patch} {
		if err := s.WriteOptimU16(u); err != nil {
			return err
		}
	}
	return nil
}

// List and Dictionary share one binary layout — both serialize a Lua table:
// u32 count, then a string key and a nested tree per entry. Lists carry
// empty keys.
func (s *Serializer) writeEntries(entries []DictEntry) error {
	if err := s.WriteU32(uint32(len(entries))); err != nil {
		return err
	}
	for _, entry := range entries {
		if err := s.WriteStrProperty(entry.Key); err != nil {
			return err
		}
		if err := s.WritePropertyTree(entry.Value); err != nil {
			return err
		}
	}
	return nil
}

// WriteList writes a property tree list, giving each item an empty key.
//
// See https://wiki.factorio.com/Property_tree#List
func (s *Serializer) WriteList(items []PropertyTree) error {
	entries := make([]DictEntry, len(items))
	for i, item := range items {
		entries[i] = DictEntry{Value: item}
	}
	return s.writeEntries(entries)
}

// WriteDictionary writes a property tree dictionary in entry order.
//
// See https://wiki.factorio.com/Property_tree#Dictionary
func (s *Serializer) WriteDictionary(entries []DictEntry) error {
	return s.writeEntries(entries)
}

// WritePropertyTree writes one property tree element.
func (s *Serializer) WritePropertyTree(pt PropertyTree) error {
	if err := s.WriteU8(uint8(pt.Kind)); err != nil {
		return err
	}
	// The "any-type" flag is unused by Factorio; always written as false.
	if err := s.WriteBool(false); err != nil {
		return err
	}

	switch pt.Kind {
	case KindNone:
		return nil
	case KindBool:
		v, ok := pt.Bool()
		if !ok {
			return valueMismatch(pt)
		}
		return s.WriteBool(v)
	case KindNumber:
		v, ok := pt.Number()
		if !ok {
			return valueMismatch(pt)
		}
		return s.WriteDouble(v)
	case KindString:
		v, ok := pt.Str()
		if !ok {
			return valueMismatch(pt)
		}
		return s.WriteStrProperty(v)
	case KindList:
		v, ok := pt.List()
		if !ok {
			return valueMismatch(pt)
		}
		return s.WriteList(v)
	case KindDict:
		v, ok := pt.Dict()
		if !ok {
			return valueMismatch(pt)
		}
		return s.WriteDictionary(v)
	case KindSignedInt:
		v, ok := pt.SignedInt()
		if !ok {
			return valueMismatch(pt)
		}
		return s.WriteLong(v)
	case KindUnsignedInt:
		v, ok := pt.UnsignedInt()
		if !ok {
			return valueMismatch(pt)
		}
		return s.WriteUnsignedLong(v)
	default:
		return &UnknownPropertyTypeError{Type: uint8(pt.Kind)}
	}
}

func valueMismatch(pt PropertyTree) error {
	return fmt.Errorf("property tree kind %v holds %T", pt.Kind, pt.Value)
}
